spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurer.java

```diff
@@ -230,17 +232,34 @@ public class RetryTopicConfigurer {
 	 * @param beanFactory the bean factory.
 	 * @param retryTopicNamesProviderFactory the retry topic names factory.
 	 */
-	@Autowired
+	@Deprecated
 	public RetryTopicConfigurer(DestinationTopicProcessor destinationTopicProcessor,
 								ListenerContainerFactoryResolver containerFactoryResolver,
 								ListenerContainerFactoryConfigurer listenerContainerFactoryConfigurer,
 								BeanFactory beanFactory,
 								RetryTopicNamesProviderFactory retryTopicNamesProviderFactory) {
 
+		this(destinationTopicProcessor, containerFactoryResolver,
+				listenerContainerFactoryConfigurer, retryTopicNamesProviderFactory);
+		this.beanFactory = beanFactory;
+	}
+
+	/**
+	 * Create an instance with the provided properties.
+	 * @param destinationTopicProcessor the destination topic processor.
+	 * @param containerFactoryResolver the container factory resolver.
+	 * @param listenerContainerFactoryConfigurer the container factory configurer.
+	 * @param retryTopicNamesProviderFactory the retry topic names factory.
+	 */
+	@Autowired
```

Comment:
Can we just don't do this?

No one is going to create a bean instance of this class without calling this ctor anyway.

From my experience all the auto-wiring features is for end-user projects: the framework stuff has to be as explicit as possible.

Reply:
We need this for the legacy bootstrapping - it's autowired there. Since we have more than one constructor, we need tell the framework which one is to be used.

An alternative would be removing the old constructor, but that would be a breaking change.

Any ideas?

Reply:
We actually used to have at least one more constructor for this class which was deprecated - probably @garyrussell removed it for 3.0, but it's probably still there for 2.x

Reply:
Right. And how it is used `new RetryTopicConfigurer()`?
So, no any auto-wire magic. Isn't it?

Reply:
We actually don't have a default constructor.

It's declared here:
https://github.com/spring-projects/spring-kafka/blob/feb35635c3066937cebfdfcb15dd2e27e17c2673/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicBootstrapper.java#L86

And used here:
https://github.com/spring-projects/spring-kafka/blob/feb35635c3066937cebfdfcb15dd2e27e17c2673/spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java#L514

The framework autowires the dependencies.

Reply:
I've just checked and the legacy constructor is still there for 2.9 and 2.8.x. So for 3.0, we can remove the now deprecated constructor and remove the `@Autowired`, but for 2.9 we'd need to remove both deprecated constructors for that.

Reply:
OK. I see:
```
registry.registerBeanDefinition(beanName, new RootBeanDefinition(beanClass))
```

So, yeah... Let's leave it for now!

Thanks

Reply:
Sure, and thanks for the pointer on making everything in the framework as explicit as possible, that’s a good thing to keep in mind.